Feature/tuya mcu update

benjaminnestler opened the pull request:
## Description:

Okay, this is the initial try to support the OTA-Upgrade of the TUYA MCU with tasmota. You can activate this feature by defining the Macro `USE_TUYA_MCU_UPGRADE`. 

There is a new TuyaCommand added `TuyaUpgrade` which needs 3 arguments (strings) to start an OTA of the tuya mcu:
- version of OTA firmware
- md5 checksum of OTA firmware file
- http-server and path to the OTA firmware file

You can start the OTA procedure via WebComand (Console) or MQTT.

Example:
`TuyaUpgrade 1.1.3,85c449b5cb2502279399dfec57ad30ac,http://192.168.178.23/16025109522a96f16f7ae.bin`

The first step is to load the OTA binary to a non-volatile memory:
- For ESP8266 the internal flash in the range of the tasmota-OTA partition is used 
- For ESP32 devices the UFS is used to store the OTA binary

In the Tuya documentation you'll find this simple (and not complete) sequence diagram
![tuya-update-sequence](https://github.com/arendst/Tasmota/assets/101095581/d1110031-6b1d-48c2-a9a1-6968462c84e6)

If OTA binary transfer succeeds, tasmota request the tuya mcu for supported update package size. MCU should response with a dedicated transfer buffer size (typically 256 bytes).

Tasmota starts to transmit the OTA binary from the source nvm to the tuya mcu. After receiving a response to an update package the next OTA file chunk will be transferred with a delay of one second.

After the full transfer of OTA file the tuya mcu should restart und tasmota can query the product ID to verify the tuya mcu version.

I don't know if there are devices out there, for which we can request a tuya mcu firmware file. Maybe a simple request in direction to the device producer leads to a pleasing response ... or not. 

Some companies using a pre-installed Tasmota version on their commercial devices. And maybe these are the companies, which are providing firmware updates for the tuya MCU.
- https://wesmartify.de/search?q=tasmota&options%5Bprefix%5D=last
- https://nous.technology/smart/tasmota/

I've two different devices (two thermostats) with tuya mcu test-firmwares to verify the OTA process on ESP32  and ESP8266 device.

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [x] The code change is tested and works with Tasmota core ESP32 V.2.0.14
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

arendst commented on tasmota/tasmota_xdrv_driver/xdrv_16_tuyamcu_upgrade.ino:
```diff
@@ -0,0 +1,390 @@
+#ifdef ESP8266
+  // prepare to use OTA-partition
+  _start = (ESP.getSketchSize() + _flash_sector_size - 1) & (~(_flash_sector_size - 1));
+  _end = (uint32_t)&_FS_start - 0x40200000;
```
Dangerous as it can overwrite ESP8266 settings located in eight banks below _FS_start. See Tasmota OTA function which makes sure the latests settings are stored safely before the Tasmota OTA is allowed to start,

benjaminnestler replied:
Fixed this by commit [c2e284b](https://github.com/arendst/Tasmota/pull/19877/commits/c2e284b565a9f56fd39f8b90a5f4d1bcb53415f8)

benjaminnestler replied:
Maybe we should take a look at FLOG::init(void) in [support_flash_log.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_support/support_flash_log.ino). That was my original source of code snippet copy to prepare the OTA part.

Jason2866 replied:
Yeah, this code part is rather old. At this time many API / functions in Tasmota just did not exists.

barbudor commented:
Tasmota OTA already includes support for upgrading other MCUs such as an Arduino (with Tasmota Client) or zigbee processors, etc...
It allows to use the standard file upload web page without the need to setup a local http server
May be path to investigate?

benjaminnestler commented:
Hi @barbudor,
I think it's a good idea! You are the pros of tasmota ;-) Thanks for your inspiration. I will check this path.

benjaminnestler commented:
Hi @barbudor,
I've added the path to upload the Tuya mcu firmware file via Webserver upload page.

barbudor commented:
> Hi @barbudor, I've added the path to upload the Tuya mcu firmware file via Webserver upload page.

Cool 👍

Jason2866 commented:
@benjaminnestler Still draft? If you want a review remove the Draft status

benjaminnestler commented:
Hi @barbudor, @Jason2866 or @arendst 

what is your best practice to refresh a pull request feature branch in sync with the current development branch?
Should I rebase the feature branch from time to time until the PR is merged?

barbudor commented:
Unless the PR and the others gets in conflict because they are touching similar parts of the code, it's not necessary.

benjaminnestler commented:
Found a bug ... after fixing it, I will reset the draft state of the PR.

benjaminnestler commented:
Hi,

Is there a good solution to debug/log Tasmota firmware in the case of an unknown source of firmware reset?

Currently, I am working on the Tuya MCU firmware upgrade process. Generally, the process runs and flashes the Tuya MCU via Tasmota. I have implemented a new TuyaMCU command to upload the Tuya MCU file and transfer this file via the Tuya protocol to the Tuya MCU.

@barbudor gave me a hint that other co-processors can be updated through the WebUI upload via the file dialog. Therefore, I want to implement this approach as well. However, following this path, I ran into a Tasmota reset before the firmware could be transferred to the Tuya MCU. I need a solution to detect the source of the reset.
If I disconnect the hardware connection (RX and TX) to the Tuya MCU, the reset doesn't occur, and the Tasmota firmware does what I want in the case of no responses from the Tuya MCU.

arendst commented:
Do you see an exception or a watchdog? The latter is likely caused by your tuya upload code taking too long in between returning to the schedular to reset the watchdog timer and more importantly keeping network active.

benjaminnestler commented:
Okay, with the help of the map file, I can pinpoint where the error might be.

benjaminnestler committed:
CHG: Seperate sourcecode in some files.

benjaminnestler committed:
FIX: ESP32/ESP8266 compatibility

benjaminnestler committed:
ADD: Check if initial upgrade package (negotiate the transfer buffer size) was responded by tuya mcu.
CHG: Refactor trigger and request flags from separate boolean to a single uint8.

benjaminnestler committed:
ADD: Initial support for TUYA MCU firmware update via Tasmota web command or MQTT

benjaminnestler committed:
FIX: Pipeline error fixed. Forgot to add a break

benjaminnestler committed:
ADD: Its better to connect the UFS usage staff to the defined USE_UFILESYS macro

benjaminnestler committed:
ADD: Support the tuya mcu file upload and upgrade process via firmware upgrade --> upload file webserver.

benjaminnestler committed:
ADD: some comments

benjaminnestler committed:
CHG: Refactor log messages, request of product ID, timeout of messages.

benjaminnestler committed:
CHG: Refactor LOG messages and type

benjaminnestler committed:
ADD: Save the tasmota config
CHG: Improve calculation of start and end addresses to upload the tuya mcu ota binary file to the (ota range of) flash

benjaminnestler committed:
ADD: Select OTA firmware file type in WebUI for different co-microcontrollers to clarify which file will be uploaded.

benjaminnestler committed:
CHG: Some tiny refactorings

benjaminnestler committed:
FIX: Mixed up source and destination pointer in memcpy_P() function

benjaminnestler committed:
CHG: Correct logs

benjaminnestler committed:
FIX: Extend the timeout after full transfer of OTA to the Tuya MCU to 30 seconds
FIX: Check if expected version is a nullptr

benjaminnestler committed:
FIX: Add success flag to see the correct upload result in WebUI after successful upload and flash process of tuya mcu

benjaminnestler committed:
FIX: Rebasing mistake fixed

benjaminnestler commented:
What else needs to be done:

I'm considering adding a progress indicator that can be shown in the WebUI during the upload and transfer to the Tuya MCU. The overall process duration depends on the OTA file size. Currently, one chunk is transferred to the Tuya MCU every second. So, the entire process takes a few minutes.

arendst commented:
>  I added a dropdown box to the WebUI to select the desired OTA upload file. 

The fact that this is currently not present is keeping the code as small as possible. Until tuya testing for differences in the first never seemed to be an issue. I know that ":" to distinguish file type isn't a secure solution but then, it was only supposed to be used for TasmoClient.

Having a drop down list still is not a secure way of uploading files; users can still select the wrong type of file.

Any way, I suggest to make the drop down list optional using a define to keep the code as small as possible.

benjaminnestler commented:
So, I've checked the source code in xdrv_01_9_webserver.ino. I can't think of a clever way to reliably distinguish between Tasmota firmware and Tuya MCU firmware. The issue is that the Tuya firmware can contain anything as the first byte. Both 0xE9 and 0x1F can occur. This makes it impossible to differentiate between the firmware to be uploaded. I need another decision criterion. It's not sufficient to only query the presence/configuration of Tuya RX and TX pins because, in this case, a Tasmota firmware update would always end up in the Tuya MCU update branch. That's why I added manual selection through the dropdown box.

Does anybody have a good idea on how to make the distinction without manual selection? @Jason2866 @barbudor @arendst

barbudor commented:
Basic question : where do you get tuya MCU firmware from? Are some vendors provide updates for download? Or are you scrapping them spying the communication on a stock device?

benjaminnestler commented:
> Looking further in the first block, is there no common tuya diferentiatorin there?

No, sorry, there is no specific marker byte used.

I've contacted two vendors and obtained their Tuya MCU firmwares. Both of them respond in the same way. There are no unique identification bytes that can be utilized.

benjaminnestler commented:
> Basic question : where do you get tuya MCU firmware from? Are some vendors provide updates for download? Or are you scrapping them spying the communication on a stock device?

Yes, indeed, I am in contact with two manufacturers and have some firmware files for OTA tests.

benjaminnestler commented:
What can I do to ensure that this branch doesn't fade into oblivion?

benjaminnestler committed:
CHG: Send firmware chunks immediately after receiving a response from the Tuya MCU for faster firmware transfer. The default behavior (without defining the macro TUYA_MCU_UPGRADE_PACKAGE_IMMEDIATELY) is to introduce a delay before sending the next firmware package.
